% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filter_ld.R
\name{filter_ld}
\alias{filter_ld}
\title{GBS/RADseq short and long distance linkage disequilibrium pruning}
\usage{
filter_ld(interactive.filter = TRUE, data, filter.short.ld = "mac",
  filter.long.ld = NULL, parallel.core = parallel::detectCores() - 1,
  filename = NULL, verbose = TRUE, ...)
}
\arguments{
\item{interactive.filter}{(optional, logical) Do you want the filtering session to
be interactive. Figures of distribution are shown before asking for filtering
thresholds.
Default: \code{interactive.filter = TRUE}.}

\item{data}{(4 options) A file or object generated by radiator:
\itemize{
\item tidy data
\item Genomic Data Structure (GDS)
}

\emph{How to get GDS and tidy data ?}
Look into \code{\link{tidy_genomic_data}},
\code{\link{write_seqarray}} or
\code{\link{tidy_vcf}}.}

\item{filter.short.ld}{(character) 5 options (default: \code{filter.short.ld = "mac"}):
\enumerate{
\item \code{filter.short.ld = "random"} for a random selection of 1 SNP on the read,
\item \code{filter.short.ld = "first"} for the first one on the read...,
\item \code{filter.short.ld = "last"} for the last SNP on the read and
\item \code{filter.short.ld = "middle"} for locus with > 2 SNPs/read the option to select at random
one SNP between the first and the last SNP on the read. If the locus as <= 2
SNPs on the read, the first one is selected. Note that for that last option,
the numbers are reported.
\item \code{filter.short.ld = "mac"} will select the SNP on the locus with the maximum global
Minor Allele Count (MAC).

Using \code{filter.short.ld = NULL}, skip this filter.
}}

\item{filter.long.ld}{(optional, double) The threshold to prune SNP based on
Long Distance Linkage Disequilibrium. The argument filter.long.ld is
the absolute value of measurement.
Default: \code{filter.long.ld = NULL}.}

\item{parallel.core}{(optional) The number of core used for parallel
execution during import.
Default: \code{parallel.core = parallel::detectCores() - 1}.}

\item{filename}{(optional, character) File name prefix for file written in
the working directory.
Default: \code{filename = NULL}.}

\item{verbose}{(optional, logical) When \code{verbose = TRUE}
the function is a little more chatty during execution.
Default: \code{verbose = TRUE}.}

\item{...}{(optional) Advance mode that allows to pass further arguments
for fine-tuning the function. Also used for legacy arguments (see details or
special section)}
}
\value{
A list in the global environment, with these objects:
\enumerate{
\item $ld.summary: tibble with LD statistics used for the boxplot
\item $ld.boxplot: box plot of LD values
\item $whitelist.ld: whitelist of markers kept after filtering for LD.
The argument \code{filter.long.ld} must be used to generate the whitelist.
\item $blacklist.ld: blacklist of markers prunned during the filtering
for LD.
The argument \code{filter.long.ld} must be used to generate the blacklist.
\item $data: The filtered tidy dataset.
\item $gds: the path to the GDS file.
}
}
\description{
SNP short and long distance linkage disequilibrium pruning.

What sets appart radiator LD pruning is the RADseq data tailored arguments:
\itemize{
\item \strong{minimize short linkage disequilibrium (LD)}:
5 values available for \code{filter.short.ld} argument (see below).

\item \strong{reduce long distance LD}:
Long distance LD pruning is usually advised to avoid capturing the variance LD
in PCA analysis.

Use the argument \code{filter.long.ld} with values between 0.7 and 0.9 is a
good starting point. Ideally, you want to visualize the LD before choosing a threshold.


Strategically, run the function with \code{filter.long.ld} argument and
refilter the data using the outlier statistic
generated by the function (printed on the figure in the output) and using
\code{long.ld.missing = TRUE}. This advanced argument will choose the best SNP
based on missing data statistics, instead of choosing randomly one SNP
(see details).
}


This function is used internally in \href{https://github.com/thierrygosselin/radiator}{radiator}
and might be of interest for users.
}
\details{
The function requires \href{https://github.com/zhengxwen/SNPRelate}{SNPRelate}
(see example below on how to install).


\strong{Advance mode, using \emph{dots-dots-dots}}
\itemize{
\item \code{maf.data} (path) this argument is no longer supported.
It's a small cost in time in favour of making sure the MAC/MAF fits the actual
data.

\item \code{long.ld.missing} (logical) With \code{long.ld.missing = TRUE}.
The function first generates long distance LD values between markers along
the same chromosome or scaffold with \emph{SNPRelate::snpgdsLDMat}.
Based on the LD threshold (\code{filter.long.ld}) SNPs in LD will be pruned
based on missingness.
e.g. if 4 SNPs are in LD, the 1 SNP selected in
the end is base on genotyping rate/missingness. If this statistic is equal
between the SNPs in LD, 1 SNP is chosen randomly.

Using missigness add extra computational time. To speed the analysis when
missingness between markers is not an issue, use \code{long.ld.missing = FALSE}.
The function will use \emph{SNPRelate::snpgdsLDpruning}
to prune the dataset. SNPs in LD are selected randomly.
Default: \code{long.ld.missing = FALSE}.
\item \code{ld.method}: (optional, character) The values available are
\code{"composite"}, for LD composite measure, \code{"r"} for R coefficient
(by EM algorithm assuming HWE, it could be negative), \code{"r2"} for r^2,
\code{"dprime"} for D',
\code{"corr"} for correlation coefficient. The method corr and composite are
equivalent when SNPs are coded based on the presence of the alternate allele
(\code{0, 1, 2}).
Default: \code{ld.method = "r2"}.

\item \code{keep.gds} (logical) Default: \code{keep.gds = FALSE}, when the input data is a
tidy data frame of genotypes, the GDS file
generated for the long distance LD is removed after completion. With GDS input,
the GDS file is always updated and kept in the working directory.
\item \code{ld.figures}: (logical) Generate long distance LD statistics and
figures.
Default: \code{ld.figures = TRUE}
\item \code{path.folder}: to write ouput in a specific path
(used internally in radiator). Default: \code{path.folder = getwd()}.
If the supplied directory doesn't exist, it's created.
}
}
\examples{
\dontrun{
require(SNPRelate)
#To install SNPRelate:
install.packages("BiocManager")
BiocManager::install ("SNPRelate")
library(radiator)
data <- radiator::tidy_vcf(data = "my.vcf", strata = "my.strata.tsv", verbose = TRUE)

# short distance LD, no long distance LD:
check.short.ld <- radiator::filter_ld(data = data$tidy.data, filter.short.ld = "mac")

# short distance LD and long distance LD:
pruned.ld <- radiator::filter_ld(
        data = data$tidy.data,
        filter.short.ld = "mac", filter.long.ld = 0.8)

# short distance LD and long distance LD, incorporating missing data:
pruned.ld <- radiator::filter_ld(
         data = data, # a GDS object generated by radiator
         filter.short.ld = "mac", filter.long.ld = 0.8, long.ld.missing = TRUE)
}
}
\references{
Zheng X, Levine D, Shen J, Gogarten SM, Laurie C, Weir BS.
(2012) A high-performance computing toolset for relatedness and principal component
analysis of SNP data. Bioinformatics. 28: 3326-3328.
doi:10.1093/bioinformatics/bts606
}
\seealso{
\href{https://github.com/zhengxwen/SNPRelate}{SNPRelate}
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
