% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filter_fis.R
\name{filter_fis}
\alias{filter_fis}
\title{Fis filter}
\usage{
filter_fis(data, approach = "haplotype", fis.min.threshold,
  fis.max.threshold, fis.diff.threshold, pop.threshold, percent, filename)
}
\arguments{
\item{data}{12 options for input: VCFs (SNPs or Haplotypes,
to make the vcf population ready),
plink, stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format. To verify that radiator detect
your file format use \code{\link{detect_genomic_format}} (see example below).
Documented in \strong{Input genomic datasets} of \code{\link{tidy_genomic_data}}.}

\item{approach}{Character. By \code{"SNP"} or by \code{"haplotype"}.
The function will consider the SNP or haplotype statistics to filter the marker.
Default by \code{"haplotype"}.}

\item{fis.min.threshold}{Number.}

\item{fis.max.threshold}{Number.}

\item{fis.diff.threshold}{Number (0 - 1)}

\item{pop.threshold}{Fixed number of pop required to keep the locus.}

\item{percent}{Is the threshold a percentage ? TRUE or FALSE.}

\item{filename}{(optional) The function uses \code{\link[fst]{write.fst}},
to write the tidy data frame in
the folder created in the working directory. The file extension appended to
the \code{filename} provided is \code{.rad}.
Default: \code{filename = NULL}.}
}
\description{
TODO
}
