% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/merge_dart.R
\name{merge_dart}
\alias{merge_dart}
\title{Merge DArT files}
\usage{
merge_dart(dart1, strata1, dart2, strata2, pop.select = NULL,
  monomorphic.out = TRUE, common.markers = TRUE, keep.rad = FALSE,
  filename = NULL, parallel.core = parallel::detectCores() - 1)
}
\arguments{
\item{dart1}{Full path of the first DArT file.}

\item{strata1}{Full path of the first strata file for dart1.}

\item{dart2}{Full path of the second DArT file.}

\item{strata2}{Full path of the second strata file for dart2.}

\item{pop.select}{(string, optional) Selected list of populations for
the analysis. e.g. \code{pop.select = c("QUE", "ONT")} to select \code{QUE}
and \code{ONT} population samples (out of 20 pops).
Default: \code{pop.select = NULL}}

\item{monomorphic.out}{(optional) Should the monomorphic
markers present in the dataset be filtered out ?
Default: \code{monomorphic.out = TRUE}.}

\item{common.markers}{(optional) Logical. Default: \code{common.markers = TRUE},
will only keep markers in common (genotyped) between all the populations.}

\item{keep.rad}{Unless the default is changed, the function removes the
temporary tidy dart files generated for each DArT datasets during import.
Default: \code{keep.rad = FALSE}.}

\item{filename}{Name of the merged DArT file.
By default, the function gives the merged data the filename:\code{merge_dart}
with date and time appended. The function will also append the date and time
to the filename provided.
The data is written in the working directory.
Default: \code{filename = NULL}.}

\item{parallel.core}{(optional) The number of core used for parallel
execution during vcf import.
Default: \code{parallel::detectCores() - 1}.}
}
\value{
The function returns a list in the global environment and 2 data frames
in the working directory. The dataframes are the tidy dataset and a strata file
of the 2 merged DArT files.
}
\description{
This function allows to merge 2 DArT files (filtered of not).
}
\details{
The function average across markers the columns: CALL_RATE, REP_AVG,
AVG_COUNT_REF and AVG_COUNT_SNP, when found in the data.
For DArT, theses columns represent:
\itemize{
\item CALL_RATE: is the proportion of samples for which the genotype was called.
\item REP_AVG: is the proportion of technical replicate assay pairs for which
the marker score is consistent.
\item AVG_COUND_REF and AVG_COUND_SNP: the coverage for the reference and
alternate alleles, respectively.
}

The function removes markers with starting with 1000 that are not immortalized by DArT


When the argument \strong{common.markers} is kept to TRUE, the function
produces an
\href{https://github.com/hms-dbmi/UpSetR}{UpSet plot} to visualize the number
of markers common or not between populations. The plot is not saved automatically,
this as to be done manually by the user.
}
\examples{
\dontrun{
# The simplest way to run the function:
sum <- radiator::merge_dart(
dart1 = "bluefin_larvae.tsv", strata1 = "strata1_bft_larvae.tsv",
dart1 = "bluefin_adults.csv", strata2 = "strata2_bft_adults.tsv",
filename = "bluefin_combined")
}
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
