% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tidy_dart.R
\name{tidy_dart}
\alias{tidy_dart}
\title{Tidy \href{http://www.diversityarrays.com}{DArT} output file.}
\usage{
tidy_dart(data, strata, verbose = FALSE,
  parallel.core = parallel::detectCores() - 1)
}
\arguments{
\item{data}{DArT output file. Note that most popular format used by DArT are
recognised (1- and 2- row format,also called binary, and count data.).
If you encounter a problem, sent me your data so that I can update
the function. The data is a tab delimited file ending with \code{.tsv}.}

\item{strata}{A tab delimited file with columns header:
\code{INDIVIDUALS} and \code{STRATA}.
Note: the column \code{STRATA} refers to any grouping of individuals.
You need to make sure that
the column \code{INDIVIDUALS} match the id used by DArT. Sometimes it's a
DArT DB integer name.
Use a third column in the strata, named \code{NEW_ID}, to replace the
\code{INDIVIDUALS} column in the main data file.}

\item{verbose}{(optional, logical) When \code{verbose = TRUE}
the function is a little more chatty during execution.
Default: \code{verbose = TRUE}.}

\item{parallel.core}{(optional) The number of core used for parallel
execution during vcf import.
Default: \code{parallel::detectCores() - 1}.}
}
\value{
A tidy dataframe with these columns:
\enumerate{
\item MARKERS: generated by radiator and correspond to CHROM + LOCUS + POS separated by 2 underscores.
\item CHROM: the chromosome, for de novo: CHROM_1.
\item LOCUS: the locus.
\item POS: the SNP id on the LOCUS.
\item REF: the reference allele.
\item ALT: the alternate allele.
\item INDIVIDUALS: the sample name.
\item POP_ID: populations id of the sample.
\item GT: the genotype in 6 digit format Ã  la genepop.
\item GT_VCF: the genotype in VCF format.
\item GT_VCF_NUC: the genotype in VCF format, but keeping the nucleotide information.
\item GT_BIN: the genotype in binary format similar to PLINK. The number correspond to the number of alternate allele in the genotype.
\item CALL_RATE: call rate output specific of DArT.
\item AVG_COUNT_REF: the coverage for the reference allele, output specific of DArT.
\item AVG_COUNT_SNP: the coverage for the alternate allele, output specific of DArT.
\item REP_AVG: the reproducibility average, output specific of DArT.
}
}
\description{
Used internally in \href{https://github.com/thierrygosselin/radiator}{radiator}
and might be of interest for users. The function generate a tidy dataset using
\href{http://www.diversityarrays.com}{DArT} file.
}
\examples{
\dontrun{
clownfish.dart.tidy <- radiator::tidy_dart(
data = "clownfish.dart.tsv",
strata = "clownfish.strata.tsv",
verbose = TRUE)
}
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com} and Peter Grewe \email{peter.grewe@csiro.au}
}
