% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tidy_vcf.R
\name{tidy_vcf}
\alias{tidy_vcf}
\title{Tidy a vcf file (bi/multi-allelic).}
\usage{
tidy_vcf(data, strata, vcf.metadata = FALSE,
  parallel.core = parallel::detectCores() - 1, verbose = FALSE, ...)
}
\arguments{
\item{data}{12 options: VCF (SNPs or Haplotypes,
to make the vcf population ready, see details below),
plink, stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format.
\emph{See details} of \code{\link[radiator]{tidy_genomic_data}}.}

\item{strata}{(optional/required) Required for VCF and haplotypes files,
optional for the other formats supported.

The strata file is a tab delimited file with a minimum of 2 columns headers:
\code{INDIVIDUALS} and \code{STRATA}.
If a \code{strata} file is specified with all file formats that don't
require it, the strata argument will have precedence on the population
groupings used internally in those file formats.
The \code{STRATA} column can be any hierarchical grouping.
To create a strata file see \code{\link[radiator]{individuals2strata}}.
If you have already run
\href{http://catchenlab.life.illinois.edu/stacks/}{stacks} on your data,
the strata file is similar to a stacks \emph{population map file},
make sure you
have the required column names (\code{INDIVIDUALS} and \code{STRATA}).
The strata column is cleaned of a white spaces that interfere with some
packages or codes: space is changed to an underscore \code{_}
Default: \code{strata = NULL}.}

\item{vcf.metadata}{(optional, logical or string) For VCF files only.
With \code{logical, TRUE/FALSE}, \code{vcf.metadata = FALSE}, only the genotype
information is kept (GT field). With \code{vcf.metadata = TRUE},
all the metadata contained in the \code{FORMAT} field will be kept in
the tidy data file. radiator is currently keeping/cleaning only these metadata:
\code{"DP", "AD", "GL", "PL", "GQ", "GOF", "NR", "NV"}.
If you need different one, submit a request.
With \code{string}, explicitely ask for a particular FORMAT field you want
to keep (except the GT field that is always imported).
e.g. \code{vcf.metadata = "PL"} or \code{vcf.metadata = c("DP", "GL")}.
Default: \code{vcf.metadata = FALSE}.}

\item{parallel.core}{(optional) The number of core used for parallel
execution during vcf import.
Default: \code{parallel::detectCores() - 1}.}

\item{verbose}{(optional, logical) When \code{verbose = TRUE}
the function is a little more chatty during execution.
Default: \code{verbose = TRUE}.}

\item{...}{(optional) To pass further argument for fine-tuning the tidying
(details below).}
}
\value{
The output in your global environment is a tidy data frame.
}
\description{
Used internally in \href{https://github.com/thierrygosselin/radiator}{radiator}
and might be of interest for users.
Highly recommended to use \code{\link[radiator]{tidy_genomic_data}} or
\code{\link[radiator]{genomic_converter}}. Those two functions allows
to manipulate and prune the dataset with blacklists and whitelists along
several other filtering options.
}
\details{
\strong{... :dot dot dot arguments}
5 arguments are available:
\enumerate{
\item whitelist.markers
\item blacklist.id
\item pop.select
\item pop.levels
\item pop.labels
}
Documentation for these arguments is detailed
in \code{\link[radiator]{tidy_genomic_data}}, the only difference here, the
arguments are objects are in the global environment, not as file in the directory.
If files are essential in your pipeline, please use \code{\link[radiator]{tidy_genomic_data}}.
}
\author{
Thierry Gosselin \email{thierrygosselin@icloud.com}
}
