<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Individual filter — filter_individual • radiator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Individual filter — filter_individual" />

<meta property="og:description" content="Filter individuals genotyped at a marker from a tidy data
set (long format) of any of these file format:
vcf, plink (tped/tfam), stacks haplotype file, genind,
genepop, data frame in wide format. The function uses
tidy_genomic_data and
tidy_wide to load the file. For filtering
you can consider the overall number of individuals (no concept of populations here),
or the number of genotyped individuals per pop. The threshold can be a fixed
number of individuals, a proportion or a percentage." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Individual filter</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/radiator/blob/master/R/radiator_deprecated_functions.R'><code>R/radiator_deprecated_functions.R</code></a></small>
    <div class="hidden name"><code>filter_individual.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Filter individuals genotyped at a marker from a tidy data
set (long format) of any of these file format:
vcf, plink (tped/tfam), stacks haplotype file, genind,
genepop, data frame in wide format. The function uses
<code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code> and
<code><a href='tidy_wide.html'>tidy_wide</a></code> to load the file. For filtering
you can consider the overall number of individuals (no concept of populations here),
or the number of genotyped individuals per pop. The threshold can be a fixed
number of individuals, a proportion or a percentage.</p>
    
    </div>

    <pre class="usage"><span class='fu'>filter_individual</span>(<span class='kw'>interactive.filter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>data</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ind.approach</span> <span class='kw'>=</span> <span class='st'>"pop"</span>, <span class='kw'>ind.threshold</span> <span class='kw'>=</span> <span class='fl'>0.5</span>, <span class='kw'>prob.pop.threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>percent</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>filename</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>interactive.filter</th>
      <td><p>(optional, logical) Do you want the filtering session to
be interactive. With default: <code>interactive.filter == TRUE</code>, the user is
asked to see figures of distribution before making decisions for filtering.</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p>12 options for input: VCFs (SNPs or Haplotypes,
to make the vcf population ready),
plink, stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format. To verify that radiator detect
your file format use <code><a href='detect_genomic_format.html'>detect_genomic_format</a></code> (see example below).
Documented in <strong>Input genomic datasets</strong> of <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>(optional)
The strata file is a tab delimited file with a minimum of 2 columns headers:
<code>INDIVIDUALS</code> and <code>STRATA</code>. Documented in <code><a href='read_strata.html'>read_strata</a></code>.
DArT data: a third column <code>TARGET_ID</code> is required.
Documented on <code><a href='read_dart.html'>read_dart</a></code>. Also use the strata read function to
blacklist individuals.
Default: <code>strata = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>ind.approach</th>
      <td><p>(character).
The approach to filter a marker, is it based on the overall number of
genotyped individuals (no concept of populations here),
<code>ind.approach = "overall"</code>, or
the number of genotyped individuals per population <code>ind.approach = "pop"</code>.
Default: <code>ind.approach = "pop"</code>.</p></td>
    </tr>
    <tr>
      <th>ind.threshold</th>
      <td><p>The individual threshold, proportion, percentage or
number e.g. 0.70, 70, 15.
Default: <code>ind.threshold = 0.5</code>.</p></td>
    </tr>
    <tr>
      <th>prob.pop.threshold</th>
      <td><p>(integer, optional) Useful to incorporate problematic
populations dragging down polymorphism discovery, but still wanted for analysis.
If individuals with missing genotypes are not managed upstream,
use this threshold to allow variance in the number of populations passing
the <code>ind.threshold</code> argument. e.g. with <code>prob.pop.threshold = 2</code>,
you tolerate a maximum of 2 populations failing to pass the
<code>ind.threshold</code>. Manage after the problematic populations
downstream with blacklist of individuals and/or missing data imputations. This
argument is not necessary with <code>ind.approach = "overall"</code>.
Default: <code>prob.pop.threshold = 0</code>. See details for more info.</p></td>
    </tr>
    <tr>
      <th>percent</th>
      <td><p>Is the threshold a percentage? TRUE or FALSE.
This argument is necessary to distinguish percentage from integer individual
threshold (e.g. 70 percent or 70 individuals).
Default: <code>percent = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional) The function uses <code><a href='http://www.fstpackage.org/reference/write_fst.html'>write.fst</a></code>,
to write the tidy data frame in
the folder created in the working directory. The file extension appended to
the <code>filename</code> provided is <code>.rad</code>.
Default: <code>filename = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) To pass further arguments for fine-tuning the function.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>With <code>interactive.filter = FALSE</code>, a list in the global environment,
with 7 objects:</p><ol>
<li><p>$tidy.filtered.ind</p></li>
<li><p>$whitelist.markers</p></li>
<li><p>$blacklist.markers</p></li>
<li><p>$strata</p></li>
<li><p>$filters.parameters</p></li>
</ol>

<p>With <code>interactive.filter = TRUE</code>, a list with 2 additionnal objects is created.</p><ol>
<li><p>$plot.ind.threshold</p></li>
<li><p>$ind.threshold.helper.table</p></li>
</ol>

<p>The object can be isolated in separate object outside the list by
following the example below.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Interactive version</strong></p>
<p>There is 2 steps in the interactive version to visualize and filter
the data based on the number of genotyped individuals:</p>
<p>Step 1. Impact of individual threshold on marker discovery</p>
<p>Step 2. Choose the filtering approach and thresholds</p>
    <p><strong>prob.pop.threshold</strong></p>
<p>If your a regular radiator user, you've seen the <code>pop.num.threshold</code>.
<code>prob.pop.threshold</code>, is a bit different and requires more thinking,
because the number of populations genotyped potentially vary across markers,
which make difficult, sometimes, the use of <code>pop.num.threshold</code>.
e.g. If only 2 populations out of 10 are genotyped for a marker and you use
<code>pop.num.threshold = 0.6</code> this could lead to inconsistensis. Thus,
it's easier to use the number of population you are willing to accept
failling to conform to the threshold.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='no'>turtle.ind</span> <span class='kw'>&lt;-</span> <span class='fu'>filter_individual</span>(
<span class='kw'>data</span> <span class='kw'>=</span> <span class='st'>"turtle.vcf"</span>,
<span class='kw'>strata</span> <span class='kw'>=</span> <span class='st'>"turtle.strata.tsv"</span>,
<span class='kw'>ind.approach</span> <span class='kw'>=</span> <span class='st'>"pop"</span>,
<span class='kw'>ind.threshold</span> <span class='kw'>=</span> <span class='fl'>50</span>,
<span class='kw'>percent</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
<span class='kw'>prob.pop.threshold</span> <span class='kw'>=</span> <span class='fl'>2</span>,
<span class='kw'>filename</span> <span class='kw'>=</span> <span class='st'>"tidy.data.turtle"</span>
)

<span class='co'>#If interactive.filter = TRUE, a list is created and to view the filtered tidy data:</span>
<span class='no'>tidy.data</span> <span class='kw'>&lt;-</span> <span class='no'>turtle.ind</span>$<span class='no'>tidy.filtered.ind</span>

<span class='co'>#Inside the same list, to isolate the blacklist.genotypes:</span>
<span class='no'>bg</span> <span class='kw'>&lt;-</span> <span class='no'>turtle.ind</span>$<span class='no'>blacklist.genotypes</span>

<span class='co'># The remaining argument are used in tidy_genomic_data during import and allow</span>
<span class='co'># the user to apply some filtering or selection before doing the filtering.</span>
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

