<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Write a <a href='https://github.com/bcm-uga/pcadapt'>pcadapt</a>
file from a tidy data frame — write_pcadapt • radiator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Write a <a href='https://github.com/bcm-uga/pcadapt'>pcadapt</a>
file from a tidy data frame — write_pcadapt" />

<meta property="og:description" content="Write a
pcadapt
file from a tidy data frame. The data is biallelic.
Used internally in
radiator
and might be of interest for users." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">radiator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="../articles/get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/life_cycle.html">Life cycle</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/radiator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Write a <a href='https://github.com/bcm-uga/pcadapt'>pcadapt</a>
file from a tidy data frame</h1>
    
    <div class="hidden name"><code>write_pcadapt.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Write a
<a href='https://github.com/bcm-uga/pcadapt'>pcadapt</a>
file from a tidy data frame. The data is biallelic.
Used internally in
<a href='https://github.com/thierrygosselin/radiator'>radiator</a>
and might be of interest for users.</p>
    
    </div>

    <pre class="usage"><span class='fu'>write_pcadapt</span>(<span class='no'>data</span>, <span class='kw'>pop.select</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>filename</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/parallel/topics/detectCores'>detectCores</a></span>() - <span class='fl'>1</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>A tidy data frame object in the global environment or
a tidy data frame in wide or long format in the working directory.
<em>How to get a tidy data frame ?</em>
Look into <span class="pkg">radiator</span> <code><a href='tidy_genomic_data.html'>tidy_genomic_data</a></code>.</p></td>
    </tr>
    <tr>
      <th>pop.select</th>
      <td><p>(optional, string) Selected list of populations for
the analysis. e.g. <code>pop.select = c("QUE", "ONT")</code> to select <code>QUE</code>
and <code>ONT</code> population samples (out of 20 pops).
Default: <code>pop.select = NULL</code></p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional) The file name prefix for the pcadapt file
written to the working directory. With default: <code>filename = NULL</code>,
the date and time is appended to <code>radiator_pcadapt_</code>.</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(optional) The number of core used for parallel
execution during import.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(optional) To pass further arguments for fine-tuning the function.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A pcadapt file is written in the working directory a genotype matrix
object is also generated in the global environment.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><em>Integrated filters:</em></p><ol>
<li><p>by defaults only markers found in common between populations are used
(See advance section).</p></li>
<li><p>by defaults monomorphic markers are automatically removed before
generating the pcadapt file.</p></li>
</ol>
    
    <h2 class="hasAnchor" id="advance-mode"><a class="anchor" href="#advance-mode"></a>Advance mode</h2>

    
    <p><em>dots-dots-dots ...</em> allows to pass several arguments for fine-tuning the function:</p><ol>
<li><p><strong>Filtering for linkage disequilibrium</strong>: 3 arguments
<code>filter.long.ld, long.ld.missing, ld.method</code>
described in <code><a href='filter_ld.html'>filter_ld</a></code> are available.
Reducing linkage before running genome scan is essential. At least start by
removing SNPs on the same RADseq locus (short linkage disequilibrium).</p></li>
<li><p><strong>Filtering markers with low Minor Allele Count</strong> : use the argument
<code>filter.mac</code> to evaluate the impact of MAC/MAF on genome scans.
The function <code><a href='filter_mac.html'>filter_mac</a></code> is called.</p></li>
<li><p>Turning off the filter that keeps markers in common between strata:
This is not recommended, but users who wants to explore the impact of such filtering
and know the biais it can potentially generate can use the argument
<code>filter.common.markers</code>.
The function <code><a href='filter_common_markers.html'>filter_common_markers</a></code> is called.
Default: <code>filter.common.markers = NULL</code></p></li>
</ol>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Luu, K., Bazin, E., &amp; Blum, M. G. (2017).
pcadapt: an R package to perform genome scans for selection based on principal component analysis.
Molecular Ecology Resources, 17(1), 67-77.</p>
<p>Duforet-Frebourg, N., Luu, K., Laval, G., Bazin, E., &amp; Blum, M. G. (2015).
Detecting genomic signatures of natural selection with principal component analysis: application to the 1000 Genomes data.
Molecular biology and evolution, msv334.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#advance-mode">Advance mode</a></li>

      <li><a href="#references">References</a></li>
          </ul>

    <h2>Author</h2>
    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Floriaan Devloo-Delva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

