# tidy_plink -------------------------------------------------------------------
#' @name tidy_plink
#' @title Tidy plink \code{.tped} and {.tfam} files in a tidy data frame
#' @description Transform a bi-allelic PLINK file in \code{.tped} format into a tidy dataset.
#' Used internally in \href{https://github.com/thierrygosselin/radiator}{radiator}
#' and \href{https://github.com/thierrygosselin/assigner}{assigner}
#' and might be of interest for users.

#' @param data The plink bi-allelic file ending with \code{.tped}, the corresponding
#' \code{.tfam} must be in the same directory.

#' @inheritParams read_strata
#' @inheritParams radiator_common_arguments

#' @references
#' PLINK: a tool set for whole-genome association and population-based linkage
#' analyses.
#' American Journal of Human Genetics. 2007: 81: 559–575. doi:10.1086/519795

#' @export
#' @rdname tidy_plink
#' @author Thierry Gosselin \email{thierrygosselin@@icloud.com}

tidy_plink <- function(data, strata = NULL, verbose = FALSE, ...) {

  # NOTE TO MYSELF: OLD CODE THAT COULD BE COMPLETELY RE_WRITTEN IF NECESSARY...
  # BUT WHO USE PLINK file as input file these days...?

  # Test
  # verbose = FALSE
  # strata = NULL
  # blacklist.id=NULL
  # pop.select=NULL
  # pop.levels=NULL
  # pop.labels=NULL


  # Function call and dotslist -------------------------------------------------
  rad.dots <- radiator_dots(
    func.name = as.list(sys.call())[[1]],
    fd = rlang::fn_fmls_names(),
    args.list = as.list(environment()),
    dotslist = rlang::dots_list(..., .homonyms = "error", .check_assign = TRUE),
    keepers = c("blacklist.id", "pop.select", "pop.levels", "pop.labels"),
    verbose = FALSE
    )


  # Strata ---------------------------------------------------------------------
  strata.df <- read_strata(
    strata = strata,
    pop.levels = pop.levels,
    pop.labels = pop.labels,
    pop.select = pop.select,
    blacklist.id = blacklist.id,
    verbose = verbose)

  # Importing file -------------------------------------------------------------
  if (verbose) message("Importing the PLINK files...")
  tfam <- data.table::fread(
    input = stringi::stri_replace_all_fixed(
      str = data,
      pattern = ".tped", replacement = ".tfam", vectorize_all = FALSE
    ),
    sep = " ",
    header = FALSE,
    stringsAsFactors = FALSE,
    verbose = FALSE,
    select = c(1,2),
    colClasses = list(character = c(1,2)),
    col.names = c("POP_ID", "INDIVIDUALS"),
    showProgress = TRUE,
    data.table = FALSE) %>%
    tibble::as_tibble(.)

  # Check if pop is the same as individuals
  if (identical(tfam$POP_ID, tfam$INDIVIDUALS)) {
    tfam %<>%
      dplyr::mutate(POP_ID = 1L)
  }

  tfam %<>%
    dplyr::mutate_at(.tbl = ., .vars = "INDIVIDUALS",
                     .funs = clean_ind_names) %>%
    dplyr::mutate_at(.tbl = ., .vars = "POP_ID",
                     .funs = clean_pop_names)



  # if no strata tfam = strata.df
  if (is.null(strata)) {
    strata.df <- strata.df <- read_strata(
      strata = tfam,
      pop.levels = pop.levels,
      pop.labels = pop.labels,
      pop.select = pop.select,
      blacklist.id = blacklist.id,
      verbose = FALSE) %$% strata
  }

  tped.header.prep <- tfam %>%
    dplyr::select(INDIVIDUALS) %>%
    dplyr::mutate(
      NUMBER = seq(1, n()),
      ALLELE1 = rep("A1", n()), ALLELE2 = rep("A2", n())
    ) %>%
    tidyr::gather(ALLELES_GROUP, ALLELES, -c(INDIVIDUALS, NUMBER)) %>%
    dplyr::arrange(NUMBER) %>%
    dplyr::select(-ALLELES_GROUP) %>%
    tidyr::unite(INDIVIDUALS_ALLELES, c(INDIVIDUALS, ALLELES), sep = "_", remove = FALSE) %>%
    dplyr::arrange(NUMBER) %>%
    dplyr::mutate(NUMBER = seq(from = (1 + 4), to = n() + 4)) %>%
    dplyr::select(-ALLELES)

  tped.header.names <- c("LOCUS", tped.header.prep$INDIVIDUALS_ALLELES)
  tped.header.integer <- c(2, tped.header.prep$NUMBER)

  # import PLINK
  input <- data.table::fread(
    input = data,
    sep = " ",
    header = FALSE,
    stringsAsFactors = FALSE,
    verbose = FALSE,
    select = tped.header.integer,
    col.names = tped.header.names,
    showProgress = TRUE,
    data.table = FALSE) %>%
    tibble::as_tibble(.) %>%
    dplyr::mutate(LOCUS = as.character(LOCUS))


  # Make tidy
  if (verbose) message("Tidying the PLINK file ...")
  # Filling GT and new separating INDIVIDUALS from ALLELES
  # combining alleles
  input <- tidyr::gather(data = input, key = INDIVIDUALS_ALLELES, value = GT, -LOCUS)

  # detect GT coding
  if (verbose) message("Scanning for PLINK tped genotype coding")
  detect.gt.coding <- unique(sample(x = input$GT, size = 100, replace = FALSE))
  gt.letters <- c("A", "C", "G", "T")

  if (TRUE %in% unique(gt.letters %in% detect.gt.coding)) {
    if (verbose) message("    genotypes coded with letters")
    gt.letters.df <- tibble::tibble(GT = c("A", "C", "G", "T", "0"), NEW_GT = c("001", "002", "003", "004", "000"))
    input  %<>% dplyr::left_join(
      gt.letters.df, by = "GT") %>%
      dplyr::select(-GT) %>%
      dplyr::rename(GT = NEW_GT)
    gt.letters.df <- NULL
  } else {
    if (verbose) message("    genotypes coded with integers")
    input  %<>%
      dplyr::mutate(GT = stringi::stri_pad_left(str = GT, width = 3, pad = "0"))
  }
  detect.gt.coding <- gt.letters <- NULL


  input  %<>%
    tidyr::separate(
      data = .,
      col = INDIVIDUALS_ALLELES,
      into = c("INDIVIDUALS", "ALLELES"),
      sep = "_") %>%
    dplyr::group_by(LOCUS, INDIVIDUALS) %>%
    tidyr::spread(data = ., key = ALLELES, value = GT) %>%
    dplyr::ungroup(.) %>%
    tidyr::unite(data = ., col = GT, A1, A2, sep = "") %>%
    dplyr::select(LOCUS, INDIVIDUALS, GT)

  # population levels and strata
  if (verbose) message("Integrating the tfam/strata file...")

  input %<>% dplyr::left_join(strata.df, by = "INDIVIDUALS")

  # removing untyped markers across all-pop
  remove.missing.gt <- input %>%
    dplyr::select(LOCUS, GT) %>%
    dplyr::filter(GT != "000000")

  untyped.markers <- dplyr::n_distinct(input$LOCUS) - dplyr::n_distinct(remove.missing.gt$LOCUS)
  if (untyped.markers > 0) {
    if (verbose) message("Number of marker with 100 % missing genotypes: ", untyped.markers)
    input <- suppressWarnings(
      dplyr::semi_join(input,
                       remove.missing.gt %>%
                         dplyr::distinct(LOCUS, .keep_all = TRUE),
                       by = "LOCUS")
    )
  }

  # Unused objects
  tped.header.prep <- tped.header.integer <- tped.header.names <- remove.missing.gt <- NULL

  # detect if biallelic give vcf style genotypes
  # biallelic <- radiator::detect_biallelic_markers(input)
  input.temp <- radiator::calibrate_alleles(data = input,
                                            verbose = verbose)
  return(res = list(input = input.temp$input, biallelic = input.temp$biallelic))
} # End import PLINK


# write_plink ------------------------------------------------------------------

#' @name write_plink
#' @title Write a plink tped/tfam file from a tidy data frame

#' @description Write a plink file from a tidy data frame.
#' Used internally in \href{https://github.com/thierrygosselin/radiator}{radiator}
#' and \href{https://github.com/thierrygosselin/assigner}{assigner}
#' and might be of interest for users.

#' @param data A tidy data frame object in the global environment or
#' a tidy data frame in wide or long format in the working directory.
#' \emph{How to get a tidy data frame ?}
#' Look into \pkg{radiator} \code{\link{tidy_genomic_data}}.

#' @param filename (optional) The file name prefix for tped/tfam files
#' written to the working directory. With default: \code{filename = NULL},
#' the date and time is appended to \code{radiator_plink_}.

#' @export
#' @rdname write_plink
#' @references Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR,
#' Bender D, et al.
#' PLINK: a tool set for whole-genome association and population-based linkage
#' analyses.
#' American Journal of Human Genetics. 2007: 81: 559–575. doi:10.1086/519795


#' @author Thierry Gosselin \email{thierrygosselin@@icloud.com}

write_plink <- function(data, filename = NULL) {

  # Import data ---------------------------------------------------------------
  if (is.vector(data)) {
    data <- radiator::tidy_wide(data = data, import.metadata = TRUE)
  }

  tped <- data %>%
    dplyr::arrange(INDIVIDUALS) %>%
    dplyr::mutate(
      COL1 = rep("0", n()),
      COL3 = rep("0", n()),
      COL4 = rep("0", n())
    ) %>%
    dplyr::select(COL1, MARKERS, COL3, COL4, INDIVIDUALS, GT) %>%
    dplyr::mutate(
      A1 = stringi::stri_sub(str = GT, from = 1, to = 3),
      A2 = stringi::stri_sub(str = GT, from = 4, to = 6)
    ) %>%
    dplyr::select(-GT) %>%
    tidyr::gather(ALLELES, GENOTYPE, -c(COL1, MARKERS, COL3, COL4, INDIVIDUALS)) %>%
    dplyr::mutate(
      GENOTYPE = as.character(as.numeric(GENOTYPE)),
      GENOTYPE = stringi::stri_pad_left(GENOTYPE, width = 2, pad = "0")
    ) %>%
    dplyr::arrange(INDIVIDUALS, ALLELES) %>%
    tidyr::unite(INDIVIDUALS_ALLELES, INDIVIDUALS, ALLELES, sep = "_") %>%
    dplyr::group_by(COL1, MARKERS, COL3, COL4) %>%
    tidyr::spread(data = ., key = INDIVIDUALS_ALLELES, value = GENOTYPE) %>%
    dplyr::arrange(MARKERS)

  tfam <- dplyr::distinct(.data = data, POP_ID, INDIVIDUALS) %>%
    dplyr::arrange(INDIVIDUALS) %>%
    dplyr::mutate(
      COL3 = rep("0",n()),
      COL4 = rep("0",n()),
      COL5 = rep("0",n()),
      COL6 = rep("-9",n())
    )

  # Create a filename to save the output files ********************************
  if (is.null(filename)) {
    # Get date and time to have unique filenaming
    file.date <- stringi::stri_replace_all_fixed(Sys.time(), pattern = " EDT", replacement = "", vectorize_all = FALSE)
    file.date <- stringi::stri_replace_all_fixed(file.date, pattern = c("-", " ", ":"), replacement = c("", "@", ""), vectorize_all = FALSE)
    file.date <- stringi::stri_sub(file.date, from = 1, to = 13)
    filename.tped <- stringi::stri_join("radiator_plink_", file.date, ".tped")
    filename.tfam <- stringi::stri_join("radiator_plink_", file.date, ".tfam")
  } else {
    filename.tped <- stringi::stri_join(filename, ".tped")
    filename.tfam <- stringi::stri_join(filename, ".tfam")
  }
  readr::write_delim(x = tped, path = filename.tped, col_names = FALSE, delim = " ")
  readr::write_delim(x = tfam, path = filename.tfam, col_names = FALSE, delim = " ")
} # end write_plink
