#' @name summary_rad
#' @title Summary statistics for RADseq data
#' @description Summarise the tidy or gds data.
#' The summary statistics is by default calculated by strata and markers.
#' Frequency of the REF/ALT alleles, the observed and the expected heterozygosity
#' and the inbreeding coefficient (FIS).
#' @param data (4 options) A file or object generated by radiator:
#' \itemize{
#' \item tidy data
#' \item Genomic Data Structure (GDS)
#' }
#'
#' \emph{How to get GDS and tidy data ?}
#' Look into \code{\link{tidy_genomic_data}},
#' \code{\link{write_seqarray}} or
#' \code{\link{tidy_vcf}}.
# @param filename (optional) Name of the file written to the working directory.
#' @inheritParams radiator_common_arguments
#' @rdname summary_rad
#' @export
# @keywords internal

summary_rad <- function(data, verbose = TRUE) {
  # Cleanup-------------------------------------------------------------------
  file.date <- format(Sys.time(), "%Y%m%d@%H%M")
  if (verbose) message("Execution date@time: ", file.date)
  old.dir <- getwd()
  opt.change <- getOption("width")
  options(width = 70)
  timing <- proc.time()# for timing
  #back to the original directory and options
  on.exit(setwd(old.dir), add = TRUE)
  on.exit(options(width = opt.change), add = TRUE)
  on.exit(timing <- proc.time() - timing, add = TRUE)
  on.exit(if (verbose) message("\nComputation time, overall: ", round(timing[[3]]), " sec"), add = TRUE)
  on.exit(if (verbose) cat("############################ summary_rad completed #############################\n"), add = TRUE)

  # Checking for missing and/or default arguments-------------------------------
  if (missing(data)) rlang::abort("Input file missing")

  # Detect format --------------------------------------------------------------
  data.type <- radiator::detect_genomic_format(data)
  if (!data.type %in% c("tbl_df", "fst.file", "SeqVarGDSClass", "gds.file")) {
    rlang::abort("Input not supported for this function: read function documentation")
  }

  if (data.type %in% c("SeqVarGDSClass", "gds.file")) {
    if (!"SeqVarTools" %in% utils::installed.packages()[,"Package"]) {
      rlang::abort('Please install SeqVarTools for this option:\n
                   install.packages("BiocManager")
                   BiocManager::install("SeqVarTools")')
    }

    if (data.type == "gds.file") {
      data <- radiator::read_rad(data, verbose = verbose)
      data.type <- "SeqVarGDSClass"
    }
  } else { #tidy.data
    if (is.vector(data)) data <- radiator::tidy_wide(data = data, import.metadata = TRUE)
  }
  s <- data %>%
    dplyr::filter(!is.na(GT_BIN)) %>%
    dplyr::group_by(MARKERS, POP_ID) %>%
    dplyr::summarise(
      N = as.numeric(n()),
      PP = as.numeric(length(GT_BIN[GT_BIN == 0])),
      PQ = as.numeric(length(GT_BIN[GT_BIN == 1])),
      QQ = as.numeric(length(GT_BIN[GT_BIN == 2]))
    ) %>%
    dplyr::mutate(
      FREQ_REF = ((PP*2) + PQ)/(2*N),
      FREQ_ALT = ((QQ*2) + PQ)/(2*N),
      HET_O = PQ/N,
      HET_E = 2 * FREQ_REF * FREQ_ALT,
      FIS = dplyr::if_else(HET_O == 0, 1, round(((HET_E - HET_O) / HET_E), 6))
    ) %>%
    dplyr::ungroup(.)

  global.maf <- s %>%
    dplyr::group_by(MARKERS) %>%
    dplyr::summarise_at(.tbl = ., .vars = c("N", "PP", "PQ", "QQ"), .funs = sum, na.rm = TRUE) %>%
    dplyr::mutate(GLOBAL_MAF = (PQ + (2 * QQ)) / (2*N)) %>%
    dplyr::select(MARKERS, GLOBAL_MAF)

  s %<>%
    dplyr::left_join(global.maf, by = "MARKERS") %>%
    dplyr::select(MARKERS, POP_ID, N, FREQ_REF, FREQ_ALT, GLOBAL_MAF, HET_O, HET_E, FIS)
  # readr::write_tsv(vcf.summary, filename, append = FALSE, col_names = TRUE)
  return(s)
}#End summary_stats_vcf_tidy
